syntax = "proto3";

package reviews;

service ReviewService {

  rpc CreateReview (CreateReviewRequest)
      returns (CreateReviewResponse);

  rpc ListReviewsByCourse (ListReviewsByCourseRequest)
      returns (ListReviewsByCourseResponse);

  rpc GetCourseRating (GetCourseRatingRequest)
      returns (GetCourseRatingResponse);

  rpc SetCourseReviewRequirement (SetCourseReviewRequirementRequest)
      returns (SetCourseReviewRequirementResponse);

  rpc GetCourseReviewRequirement (GetCourseReviewRequirementRequest)
      returns (GetCourseReviewRequirementResponse);
}

message CreateReviewRequest {
  string course_id = 1;
  string user_id = 2;

  int32 rating = 3;
  string comment = 4;

  repeated string tags = 5;
}

message ListReviewsByCourseRequest {
  string course_id = 1;

  int32 limit = 2;
  int32 offset = 3;
}

message GetCourseRatingRequest {
  string course_id = 1;
}

message SetCourseReviewRequirementRequest {
  string course_id = 1;

  int32 required_reviews_count = 2;
}

message GetCourseReviewRequirementRequest {
  string course_id = 1;
}

message CreateReviewResponse {
  Review review = 1;
}

message ListReviewsByCourseResponse {
  repeated Review items = 1;
  int32 total = 2;
}

message GetCourseRatingResponse {
  string course_id = 1;

  float average_rating = 2;
  int32 reviews_count = 3;
}

message SetCourseReviewRequirementResponse {
  CourseReviewRequirement requirement = 1;
}

message GetCourseReviewRequirementResponse {
  CourseReviewRequirement requirement = 1;
}

message Review {
  string id = 1;

  string course_id = 2;
  string user_id = 3;

  int32 rating = 4;
  string comment = 5;

  repeated string tags = 6;

  int64 created_at = 8;
}

message CourseReviewRequirement {
  string course_id = 1;

  int32 required_reviews_count = 2;
  int32 current_reviews_count = 3;

  bool is_satisfied = 4;
}
